---
title: "Macosko et al 2015 Mouse Retina"
output: html_document
---

```{r}
suppressPackageStartupMessages(library(SingleCellExperiment))
library(Matrix)
source("./real/util/data_loading.R")
fp<-file.path
bp<-"./real/macosko_2015"
```

# Overview 
This is the code to produce a *SingleCellExperiment* object of the mouse retina dropseq scRNA-Seq experiment by Macosko et al. (2015) and GEO accession [GSE63472](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE63472). See also their website http://mccarrolllab.org/dropseq/

There are 7 samples (batches). Description extracted from GEO: 
Summary: A cell suspension was prepared from wild-type P14 mouse retinas, and single-cell mRNAseq libraries were generated with Drop-Seq.

Overall design:	Drop-Seq was performed on four separate days using the same age (P14) and strain (C57BL/6). On day 1, replicate 1 was obtained. On day 2, replicates 2 and 3 were obtained. On day 3, replicates 4-6 were obtained. On day 4, replicate 7 was obtained.

# Download Cell IDs and Cluster Assignments

Use these to whitelist cell barcodes later on.
```{bash}
pushd real/macosko_2015
wget http://mccarrolllab.org/wp-content/uploads/2015/05/retina_clusteridentities.txt -P extdata
```
separate out whitelist of valid barcodes
```{r}
bc_wl<-fread(fp(bp,"extdata/retina_clusteridentities.txt"))$V1
bc_wl<-strsplit(bc_wl,"_",fixed=TRUE)
bc_wl<-sapply(bc_wl,function(x){x[2]})
write(bc_wl,file=fp(bp,"extdata/barcodes.txt"))
```

# Getting FASTQ files from SRA

The information which connects the sample information from GEO with the SRA run id is downloaded from [SRA](https://www.ncbi.nlm.nih.gov/sra?term=SRP050054) using the **Send to: File button**.

```{r}
spp<-list("Homo_sapiens_GRCh38.idx","Mus_musculus_GRCm38.idx")
names(spp)<-c("Homo sapiens","Mus musculus")

parse_sra_csv(bp,"DropSeq",spp) #from ../util/data_loading.R
```

refer to `../util/preprocessing.Rmd` for how to download the files, convert to fastq format, pseudoalign with kallisto, and process the resulting bus files so they can be parsed into a counts matrix.

# Processing FASTQs into BUS files with Kallisto

Refer to *../util/preprocessing.Rmd* for how to download the files, convert to fastq format, pseudoalign with kallisto, and process the resulting bus files with bustools.

# Load gene counts into sparse matrix

```{r}
coldata<-fread(fp(bp,"extdata/coldata.txt"))
coldata<-coldata[,c("geo_accession","Run","ScientificName")]
coldata<-coldata[order(coldata$Run),]
coldata$sample<-NA
cm<-fread(fp(bp,"extdata/retina_clusteridentities.txt"))
colnames(cm)<-c("id","cluster")
cm$bc<-scan(fp(bp,"extdata/barcodes.txt"),character())
cm$sample<-sapply(strsplit(cm$id,"_",fixed=TRUE),function(x){x[1]})

#figure out which sample ID corresponds to which SRA ID
smps<-c("p1",paste0("r",1:6))

dat<-list()
for(i in coldata$Run){
  coo<-fread(fp(bp,"data/original/genecounts",paste0(i,".txt")))
  bc<-unique(coo$bc)
  for(j in smps){
    bcref<-cm$bc[cm$sample==j]
    if(length(setdiff(bcref,bc))==0){
      coldata$sample[coldata$Run==i]<-j
      dat[[i]]<-coo[bc %in% bcref,]
      break
    }
  }
}
for(i in coldata$Run){
  smp<-coldata$sample[coldata$Run==i]
  dat[[i]]$sample<-smp
  dat[[i]]$id<-paste0(dat[[i]]$sample,"_",dat[[i]]$bc)
}
dat2<-rbindlist(dat)
dat2$gene<-factor(dat2$gene); dat2$id<-factor(dat2$id)
i<-as.integer(dat2$gene); j<-as.integer(dat2$id)
umi<-sparseMatrix(i=i,j=j,x=dat2$umi_ct)
rd<-sparseMatrix(i=i,j=j,x=dat2$read_ct)
rownames(umi)<-rownames(rd)<-levels(dat2$gene)
colnames(umi)<-colnames(rd)<-levels(dat2$id)
cm2<-merge(cm,coldata,by="sample")
cm2<-as.data.frame(cm2)
rownames(cm2)<-cm2$id
cm2<-cm2[colnames(umi),]
sce<-SingleCellExperiment(assays=list(counts=umi,read_counts=rd),colData=cm2)
saveRDS(sce,file=fp(bp,"data/00_sce_all_genes_all_cells_unrounded.rds"))

umi2<-round(umi)
gg<-rowSums(umi2)>0
rd2<-round(rd)
sce2<-SingleCellExperiment(assays=list(counts=umi2[gg,],read_counts=rd2[gg,]),colData=cm2)
saveRDS(sce2,file=fp(bp,"data/01_sce_all_genes_all_cells.rds"))
```

create a subset of data

```{r}
#sce2<-readRDS(fp(bp,"data/01_sce_all_genes_all_cells.rds"))
cm<-colData(sce2)
sce3<-sce2[,cm$sample=="r6"]
gg<-rowSums(counts(sce3))>0
sce3<-sce3[gg,]
saveRDS(sce3,fp(bp,"data/02_sce_subset_r6.rds"))
```

# Session information

```{r}
sessionInfo()
```
